curvefs/src/mds/fs_manager.cpp

```diff
@@ -0,0 +1,424 @@
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        return ret;
+    }
+
+    // 4. use metaserver interface, insert rootinode
+    uint32_t fsId = newFsInfo.GetFsId();
+    uint32_t uid = 0;  // TODO(cw123)
+    uint32_t gid = 0;  // TODO(cw123)
+    uint32_t mode = S_IFDIR | 0777;  // TODO(cw123)
+    ret = metaserverClient_->CreateRootInode(fsId, uid, gid, mode);
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "CreateFs fail, insert root inode fail"
+                   << ", fsName = " << fsName
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        FSStatusCode ret2 = fsStorage_->Delete(fsName);
+        if (ret2 != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, insert root inode fail,"
+                       << " then delete fs fail, fsName = " << fsName
+                        << ", errCode = " << ret2;
+            return ret2;
+        }
+        return ret;
+    }
+
+    // 5. update fs status to INITED
+    newFsInfo.SetStatus(FsStatus::INITED);
+    ret = fsStorage_->Update(newFsInfo);
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                   << ", fsName = " << fsName
+                   << ", blockSize = " << blockSize
+                   << ", ret = " << FSStatusCode_Name(ret);
+        ret = metaserverClient_->DeleteInode(fsId, GetRootId());
+        if (ret != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, update fs status to inited fail"
+                       << ", then delete root inode fail"
+                        << ", fsName = " << fsName
+                        << ", blockSize = " << blockSize
+                        << ", ret = " << FSStatusCode_Name(ret);
+            return ret;
+        }
+
+        ret = fsStorage_->Delete(fsName);
+        if (ret != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, insert root inode fail, "
+                       << "then delete fs fail, fsName = " << fsName
+                       << ", errCode = " << FSStatusCode_Name(ret);
+        }
+        return ret;
+    }
+
+    newFsInfo.ConvertToProto(fsInfo);
+    return FSStatusCode::OK;
+}
+
+FSStatusCode FsManager::CreateFs(const std::string& fsName, uint64_t blockSize,
+                        const S3Info& s3Info, FsInfo* fsInfo) {
+    // 1. query fs
+    if (fsStorage_->Exist(fsName)) {
+        LOG(WARNING) << "CreateFs fail, fs exist, fsName = " << fsName
+                  << ", blockSize = " << blockSize;
+        return FSStatusCode::FS_EXIST;
+    }
+
+    // 2. create fs
+    uint64_t fsSize = std::numeric_limits<uint64_t>::max();
+    MdsFsInfo newFsInfo(GetNextFsId(), fsName, FsStatus::NEW, GetRootId(),
+                            fsSize, blockSize, s3Info);
+
+    // 3. insert fs, fs status is NEW
+    FSStatusCode ret = fsStorage_->Insert(newFsInfo);
```

Comment:
reuse the same logical through a new protected member function

Reply:
done